components/script/dom/htmlelement.rs

```diff
@@ -421,11 +424,62 @@ impl HTMLElementMethods for HTMLElement {
     }
 
     // https://html.spec.whatwg.org/multipage/#the-innertext-idl-attribute
-    fn SetInnerText(&self, _: DOMString) {
-        // XXX (ferjm) implement this.
+    fn SetInnerText(&self, input: DOMString) {
+        // Step 1.
+        let document = document_from_node(self);
+
+        // Step 2.
+        let fragment = DocumentFragment::new(&document);
+
+        // Step 3. The given value is already named 'input'.
+
+        // Step 4.
+        let mut position = input.chars().peekable();
+
+        // Step 5.
+        let mut text = String::new();
+
+        // Step 6.
+        while let Some(ch) = position.next() {
+            match ch {
+                '\u{000A}' | '\u{000D}' => {
+                    if ch == '\u{000D}' && position.peek() == Some(&'\u{000A}') {
+                        // a \r\n pair should only generate one <br>,
+                        // so just skip the \r.
+                        position.next();
+                    }
+
+                    if !text.is_empty() {
+                        append_text_node_to_fragment(&document, &fragment, text);
+                        text = String::new();
+                    }
+
+                    let br = HTMLBRElement::new(local_name!("br"), None, &document);
+                    fragment.upcast::<Node>().AppendChild(&br.upcast()).unwrap();
+                },
+                _ => {
+                    text.push(ch);
+                }
+            }
+        }
+
+        if !text.is_empty() {
+            append_text_node_to_fragment(&document, &fragment, text);
+        }
+
+        // Step 7.
+        Node::replace_all(Some(fragment.upcast()), self.upcast::<Node>());
     }
 }
 
+fn append_text_node_to_fragment(document: &Document,
```

Comment:
nit: I think the standard indentation now is:

```rust
fn append_text_node_to_fragment(
    document: &Document,
    fragment: &DocumentFragment,
    text: String,
) {
    ...
}

Reply:
Sure. `DOMString` wraps a `String`. Thus, `DOMString::from(String)` doesn't need to copy `text`, while `DOMString::from(&str)` does.